File: analyzer.py
import time

from stringdata import get_data

def main():
    string1 = 'not_here'
    string2 = 'mzzzz'
    string3 = 'aaaaa'

    container = get_data()

    start = time.time()
    index = linear_search(container,string1)
    stop = time.time()
    print("time for linear search for not_here:", (stop-start) , "index found at:", index)

    start = time.time()
    index = binary_search(container, string1)
    stop = time.time()
    print("time for binary search for not_here:", (stop-start),"index found at:", index)

    start  = time.time()
    index = linear_search(container, string2)
    stop = time.time()
    print("time for linear search for mzzzz:", (stop-start),"index found at:", index)

    start = time.time()
    index = binary_search(container, string2)
    stop = time.time()
    print("time for binary search for mzzzz:", (stop-start),"index found at:", index)

    start = time.time()
    index = linear_search(container, string3)
    stop = time.time()
    print("time for linear search for aaaaa:", (stop-start),"index found at:", index)

    start = time.time()
    index = binary_search(container, string3)
    stop = time.time()
    print("time for linear search for aaaaa:",(stop-start),"index found at:", index)





def linear_search(container: tuple, element: str):
    for j in range(len(container)):
        if (container[j] == element):
            return j

    return -1

def binary_search(container: tuple,element:str):
    count =0
    high = int(len(container))
    low=0
    done = False
    prevMid=-1
    while (done==False):
        middle = low + (high-low)//2
        if(container[middle] == element):
            return middle
        elif(container[middle]<element):
            if(middle+1)<len(container):
                low = middle+1
            else:
                return False


        elif(container[middle]>element):
            if((middle-1)>=0):
                high = middle-1
            else:
                return False
        if(middle==prevMid):
            return -1
        prevMid= middle

    return -1





if __name__ == '__main__':
    main()

File: analyzer.cpp
#include <iostream>
using namespace std;
#include <vector>
#include <string>
#include <fstream>
#include "StringData.h"






int linear_search(vector<string> container, string element) {
    int size = container.size();
    for (int i =0; i<size; i++) {
        if(container[i] == element) {
            return i;
        }
    }
    return -1;

}



int binary_search(vector<string> container, string element) {
    int count = 0;
    int high = container.size();
    int low = 0;
    bool done = false;
    int prevMid =-1;
    while(done==false) {
        int middle = low+((high-low)/2);
        if(container[middle] == element) {
            return middle;
        }
        else if(container[middle]<element) {
            if((middle+1)<container.size()) {
                low = middle+1;
            }

        }
        else if(container[middle]>element) {
            if((middle-1)>-1) {
                high = middle-1;
            }
        }
        if(middle==prevMid) {
            return -1;
        }
        prevMid = middle;
    }
    return -1;

}


int main()
{

    vector<string> test = getStringData();
    string test1 = "aaaaa";
    string test2 = "mzzzz";
    string test3 = "not_here";
    double num1=0;
    double num2=0;
     num1 = systemTimeNanoseconds();
    cout << "Index:"<<binary_search(test,test1) << endl;
    num2 = systemTimeNanoseconds();
    cout<< "Binary search time for aaaaa: "<<num2-num1<<endl;

    num1 = systemTimeNanoseconds();
    cout << "Index:"<<linear_search(test,test1) << endl;
    num2 = systemTimeNanoseconds();
    cout<< "Linear search time for aaaaa: "<<num2-num1<<endl;

    num1 = systemTimeNanoseconds();
    cout << "Index:"<<binary_search(test,test2) << endl;
    num2 = systemTimeNanoseconds();
    cout<< "Binary search time for mzzzz: "<<num2-num1<<endl;

    num1 = systemTimeNanoseconds();
    cout << "Index:"<<linear_search(test,test2) << endl;
    num2 = systemTimeNanoseconds();
    cout<< "Linear search time for mzzzz: "<<num2-num1<<endl;

    num1 = systemTimeNanoseconds();
    cout << "Index:"<<binary_search(test,test3) << endl;
    num2 = systemTimeNanoseconds();
    cout<< "Binary search time for not_here: "<<num2-num1<<endl;

    num1 = systemTimeNanoseconds();
    cout << "Index:"<<linear_search(test,test3) << endl;
    num2 = systemTimeNanoseconds();
    cout<< "Linear search time for not_here: "<<num2-num1<<endl;


    return 0;

}
